ownCloud/Settings/Passcode/AppLockManager.swift

```diff
@@ -0,0 +1,256 @@
+
+	private var userDefaults: UserDefaults
+
+	// MARK: - State
+	var lastApplicationBackgroundedDate : Date?
+
+	private var failedPasscodeAttempts: Int {
+		get {
+			return userDefaults.integer(forKey: "applock-failed-passcode-attempts")
+		}
+		set(newValue) {
+			self.userDefaults.set(newValue, forKey: "applock-failed-passcode-attempts")
+		}
+	}
+	private var lockedUntilDate: Date? {
+		get {
+			return userDefaults.object(forKey: "applock-locked-until-date") as? Date
+		}
+		set(newValue) {
+			self.userDefaults.set(newValue, forKey: "applock-locked-until-date")
+		}
+	}
+
+	private let maximumPasscodeAttempts: Int = 3
+	private let powBaseDelay: Double = 1.5
+	private var lockTimer: Timer?
+
+	// MARK: - Passcode
+	private let keychainAccount = "app.passcode"
+	private let keychainPasscodePath = "passcode"
+	private let keychainLockEnabledPath = "lockEnabled"
+
+	private var keychain : OCKeychain? {
+		return OCAppIdentity.shared().keychain
+	}
+
+	var passcode: String? {
+		get {
+			if let passcodeData = self.keychain?.readDataFromKeychainItem(forAccount: keychainAccount, path: keychainPasscodePath) {
+				return String(data: passcodeData, encoding: .utf8)
+			}
+
+			return nil
+		}
+
+		set(newPasscode) {
+			if let passcode = newPasscode {
+				_ = self.keychain?.write(passcode.data(using: .utf8), toKeychainItemForAccount: keychainAccount, path: keychainPasscodePath)
+			} else {
+				_ = self.keychain?.removeItem(forAccount: keychainAccount, path: keychainPasscodePath)
+			}
+		}
+	}
+
+	// MARK: - Settings
+	var lockEnabled: Bool {
+		get {
+			return userDefaults.bool(forKey: "applock-lock-enabled")
+		}
+		set(newValue) {
+			self.userDefaults.set(newValue, forKey: "applock-lock-enabled")
+		}
+	}
+
+	var lockDelay: Int {
+		get {
+			return userDefaults.integer(forKey: "applock-lock-delay")
+		}
+
+		set(newValue) {
+			self.userDefaults.set(newValue, forKey: "applock-lock-delay")
+		}
+	}
+
+	// MARK: - Init
+	static var shared = AppLockManager()
+
+	public override init() {
+		userDefaults = OCAppIdentity.shared().userDefaults
+
+		super.init()
+
+		NotificationCenter.default.addObserver(self, selector: #selector(self.appDidEnterBackground), name: .UIApplicationDidEnterBackground, object: nil)
+		NotificationCenter.default.addObserver(self, selector: #selector(self.appWillEnterForeground), name: .UIApplicationWillEnterForeground, object: nil)
+	}
+
+	deinit {
+		NotificationCenter.default.removeObserver(self, name: .UIApplicationDidEnterBackground, object: nil)
+		NotificationCenter.default.removeObserver(self, name: .UIApplicationWillEnterForeground, object: nil)
+	}
+
+	// MARK: - Show / Dismiss Passcode View
+	func showLockscreenIfNeeded(forceShow: Bool = false) {
+		if self.shouldDisplayLockscreen || forceShow {
+			if passcodeViewController == nil {
+				passcodeViewController = PasscodeViewController(cancelHandler: {}, passcodeCompleteHandler: { (passcode: String) in
```

Comment:
You can code the initializer with `cancelHandler` as an optional so you don't need to pass it `cancelHandler: {}` and you can ignore it.

Reply:
Done 👍

Reply:
Thanks for pointing me to this. I also added a default value, so it doesn't need to be provided. Also renamed `passcodeCompleteHandler` to `completionHandler`.